core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -470,6 +474,99 @@ public long contentLength() {
+        final List<MediaType> cached = (List<MediaType>) cache.get(HttpHeaderNames.ACCEPT);
+        if (cached != null) {
+            if (cached instanceof ImmutableList) {
+                return cached;
+            } else {
+                // Should return an immutable list.
+                final List<MediaType> immutableCache = ImmutableList.copyOf(cached);
+                cache.put(HttpHeaderNames.ACCEPT, immutableCache);
+                return immutableCache;
+            }
+        }
+
+        final List<String> acceptHeaders = getAll(HttpHeaderNames.ACCEPT);
+        if (acceptHeaders.isEmpty()) {
+            cache.put(HttpHeaderNames.ACCEPT, ImmutableList.of());
+            return ImmutableList.of();
+        }
+
+        // Cache miss. Parse Accept headers and store it to cache.
+        final List<MediaType> acceptTypes = new ArrayList<>(4);
+        for (final String acceptHeader : acceptHeaders) {
+            for (String accept : ACCEPT_SPLITTER.split(acceptHeader)) {
+                try {
+                    acceptTypes.add(MediaType.parse(accept));
+                } catch (IllegalArgumentException ignored) {
+                    // Ignore a malformed MediaType.
+                }
+            }
+        }
+
+        if (acceptTypes.size() > 1) {
+            acceptTypes.sort(HttpHeadersBase::compareMediaType);
+        }
+        final List<MediaType> parsed = ImmutableList.copyOf(acceptTypes);
+        cache.put(HttpHeaderNames.ACCEPT, parsed);
+        return parsed;
+    }
+
+    final void accept(Iterable<MediaType> newAcceptTypes) {
+        requireNonNull(newAcceptTypes, "newAcceptTypes");
+        final int size = Iterables.size(newAcceptTypes);
+        checkArgument(size > 0, "newAcceptTypes is empty");
+
+        @SuppressWarnings("unchecked")
+        List<MediaType> cachedAcceptTypes = (List<MediaType>) cache.get(HttpHeaderNames.ACCEPT);
+        if (cachedAcceptTypes == null) {
+            // Set new Accept headers
+            final List<MediaType> copied = new ArrayList<>(size + 2);
+            for (MediaType mediaType : newAcceptTypes) {
+                copied.add(mediaType);
+            }
+            cachedAcceptTypes = copied;
+        } else {
+            if (cachedAcceptTypes instanceof ImmutableList) {
+                // Make cached MediaTypes mutable
+                cachedAcceptTypes = new ArrayList<>(cachedAcceptTypes);
+            }
+
+            assert cachedAcceptTypes instanceof ArrayList;
+            final ArrayList<MediaType> mediaTypeList = (ArrayList<MediaType>) cachedAcceptTypes;
+            // Merge old AcceptTypes with new one.
+            for (MediaType newAcceptType : newAcceptTypes) {
+                mediaTypeList.add(newAcceptType);
+            }
+        }
+
+        if (cachedAcceptTypes.size() > 1) {
+            cachedAcceptTypes.sort(HttpHeadersBase::compareMediaType);
+        }
+
+        cache.put(HttpHeaderNames.ACCEPT, cachedAcceptTypes);
```

Comment:
How about making the `cachedAcceptTypes` immutable before we pushing it?

Reply:
I used the same approach used in `Cookies`. https://github.com/line/armeria/pull/3714#discussion_r678065439 That would be a bit efficient if users set accept headers like:
```java
RequestHeaders.builder()
              .accept(MediaType.JSON)
              .accept(MediaType.PLAIN_TEXT)
              ...
```

Reply:
Ah, thanks. I haven't thought about it. 😄